src/h3lib/lib/algos.c

```diff
@@ -677,36 +673,207 @@ void H3_EXPORT(polyfill)(const GeoPolygon* geoPolygon, int res, H3Index* out) {
+    // them to the search hash. The hexagon containing the geofence point may or
+    // may not be contained by the geofence (as the hexagon's center point may
+    // be outside of the boundary.)
+    for (int i = 0; i < geofence.numVerts; i++) {
+        GeoCoord origin, destination;
+        origin = geofence.verts[i];
+        destination = i == geofence.numVerts - 1 ? geofence.verts[0]
+                                                 : geofence.verts[i + 1];
+        const int numHexesEstimate =
+            lineHexEstimate(&origin, &destination, res);
+        for (int j = 0; j < numHexesEstimate; j++) {
+            GeoCoord interpolate;
+            interpolate.lat =
+                (origin.lat * (numHexesEstimate - j) / numHexesEstimate) +
+                (destination.lat * j / numHexesEstimate);
+            interpolate.lon =
+                (origin.lon * (numHexesEstimate - j) / numHexesEstimate) +
+                (destination.lon * j / numHexesEstimate);
+            H3Index pointHex = H3_EXPORT(geoToH3)(&interpolate, res);
+            // A simple hash to store the hexagon, or move to another place if
+            // needed
+            int loc = (int)(pointHex % numHexagons);
+            int loopCount = 0;
+            while (found[loc] != 0) {
+                if (loopCount > numHexagons) {
+                    free(search);
+                    free(found);
+                    free(bboxes);
+                    return -1;
+                }
+                if (found[loc] == pointHex)
+                    break;  // At least two points of the geofence are identical
+                loc = (loc + 1) % numHexagons;
+                loopCount++;
+            }
+            if (found[loc] == pointHex)
+                continue;  // Skip this hex, already exists in the found hash
+            // Otherwise, set it in the found hash for now
+            found[loc] = pointHex;
+
+            search[numSearchHexes] = pointHex;
+            numSearchHexes++;
         }
-        // Check if hexagon is inside of polygon
-        GeoCoord hexCenter;
-        H3_EXPORT(h3ToGeo)(out[i], &hexCenter);
-        hexCenter.lat = constrainLat(hexCenter.lat);
-        hexCenter.lon = constrainLng(hexCenter.lon);
-        // And remove from list if not
-        if (!pointInsidePolygon(geoPolygon, bboxes, &hexCenter)) {
-            out[i] = H3_INVALID_INDEX;
+    }
+
+    // 2. Iterate over all holes, get all points of the holes, convert them to
+    // hexagons and add to only the search hash. We're going to temporarily use
+    // the `found` hash to use for dedupe purposes and then re-zero it once
+    // we're done here, otherwise we'd have to scan the whole set on each insert
+    // to make sure there's no duplicates, which is very inefficient.
+    for (int i = 0; i < geoPolygon->numHoles; i++) {
+        Geofence* hole = &(geoPolygon->holes[i]);
+        for (int j = 0; j < hole->numVerts; j++) {
```

Comment:
Could we break the body of this loop into a function call and use it here and for the outer outline above? This shouldn't have any perf effect in most cases, as it would only be called once for each loop. Seems like we could DRY out ~40 LOC

Reply:
That seems perfectly reasonable.